File: src/main/java/com/github/nsc/de/shake/lexer/characterinputstream/charactersource/CharArraySource.java
package com.github.nsc.de.shake.lexer.characterinputstream.charactersource;

class CharArraySource implements CharacterSource {

    private final char[] chars;
    private final String location;

    public CharArraySource(char[] chars, String location) {
        this.chars = chars;
        this.location = location;
    }

    @Override
    public char[] get(int start, int end) {
        char[] chars = new char[end - start];
        int length = end - start;
        if (length >= 0) System.arraycopy(this.chars, start, chars, 0, length);
        return chars;
    }

    @Override
    public char[] getAll() {
        return this.chars;
    }

    @Override
    public int getLength() {
        return this.chars.length;
    }

    @Override
    public String getLocation() {
        return location;
    }

}

File: lexer/src/main/kotlin/com/github/nsc/de/shake/lexer/characterinput/charactersource/CharArraySource.kt
package com.github.nsc.de.shake.lexer.characterinput.charactersource

internal class CharArraySource(

    override val all: CharArray,
    override val location: String?

) : CharacterSource {

    override val length: Int
        get() = all.size

    override fun get(start: Int, end: Int): CharArray {
        val chars = CharArray(end - start)
        val length = end - start
        if (length >= 0) System.arraycopy(all, start, chars, 0, length)
        return chars
    }
}